_posts/2020-10-24-text-layout.md

```diff
@@ -0,0 +1,225 @@
+categories: [text]
+---
+I love text layout, and have been working with it in one form or other for over 35 years. Yet, knowledge about it is quite arcane. I don't believe there is a single place where it's all properly written down. I have some explanation for that: while basic text layout is very important for UI, games, and other contexts, a lot of the "professional" needs around text layout are embedded in *much* more complicated systems such as Microsoft Word or a modern Web browser.
+
+A complete account of text layout would be at least a small book. Since there's no way I can write that now, this blog post is a small step towards that – in particular, an attempt to describe the "big picture," using the conceptual framework of a "loose hierarchy." Essentially, a text layout engine breaks the input into finer and finer grains, then reassembles the results into a text layout object suitable for drawing, measurement, and hit testing.
+
+The main hierarchy is concerned with laying out the entire paragraph as a single line of text. Line breaking is also important, but has a separate, parallel hierarchy.
+
+## The main text layout hierarchy
+
+The hierarchy is: paragraph segmentation as the coarsest granularity, followed by rich text style and BiDi analysis, then itemization (coverage by font), then Unicode script, and shaping clusters as the finest.
+
+![diagram of layout hierarchy](/assets/layout_pyramid.svg)
+
+### Paragraph segmentation
+
+The coarsest, and also simplest, segmentation task is paragraph segmentation. Most of the time, paragraphs are simply separated by newline (U+000A) characters, though Unicode in its infinite wisdom specifies a number of code point sequences that function as paragraph separators in plain text:
+
+* U+000A LINE FEED
+* U+000B VERTICAL TAB
+* U+000C FORM FEED
+* U+000D CARRIAGE RETURN
+* U+000D U+000A (CR + LF)
+* U+0085 NEXT LINE
+* U+2008 LINE SEPARATOR
+* U+2009 PARAGRAPH SEPARATOR
+
+In rich text, paragraphs are usually indicated through markup rather than special characters, for example `<p>` or `<br>` in HTML. But in this post, as in most text layout APIs, we'll treat rich text as plain text + attribute spans.
+
+### Rich text style
+
+A paragraph of rich text may contain *spans* that can affect formatting. In particular, choice of font, font weight, italic or no, and a number of other attributes can affect text layout. Thus, each paragraph is typically broken into a some number of *style runs,* so that within a run the style is consistent.
+
+Note that some style changes don't *necessarily* affect text layout. A classic example is color. Firefox, rather famously, does *not* define segmentation boundaries here for color changes. If a color boundary cuts a ligature, it uses fancy graphics techiques to render parts of the ligature in different color. But this is a subtle refinement and I think not required for basic text rendering. For more details, see [Text Rendering Hates You].
+
```

Comment:
That is not very subtle as it seems, it might seem fancy for Latin ligatures, but breaking runs at color and other style changes (underline, overline, shadow, etc.) would break basic shaping of many scripts rendering the text unreadable.